Add unit tests for System.Drawing.Imaging.Metafile based on Mono's unit tests

qmfrederik opened the pull request:
- Import tests for the `Metafile` class based on Mono's unit test

qmfrederik commented:
These classes currently have no code coverage.

qmfrederik commented:
@dotnet-bot test Linux x64 Release Build please
@dotnet-bot test OSX x64 Debug Build please
@dotnet-bot test UWP CoreCLR x64 Debug Build please
@dotnet-bot test Windows x64 Debug Build please
@dotnet-bot test Windows x86 Release Build please

qmfrederik commented:
All the above builds timed out, not sure why.

mellinoe commented on src/System.Drawing.Common/tests/mono/System.Imaging/MetafileTest.cs:
```diff
@@ -0,0 +1,705 @@
+        }
+
+        [ConditionalFact(Helpers.GdiplusIsAvailable)]
+        public void Metafile_String_Null()
```
Dupe of Ctor_NullString_ThrowsArgumentNullException -- we should enable that one and get it to work if this test is passing.

mellinoe commented on src/System.Drawing.Common/tests/mono/System.Imaging/MetafileTest.cs:
```diff
@@ -0,0 +1,705 @@
+        [ConditionalFact(Helpers.GdiplusIsAvailable)]
+        public void Metafile_Stream_Null()
+        {
+            AssertExtensions.Throws<ArgumentNullException, ArgumentException>("stream", null, () => new Metafile((Stream)null));
```
Dupe of Ctor_NullStream_ThrowsArgumentException

mellinoe commented on src/System.Drawing.Common/tests/mono/System.Imaging/MetafileTest.cs:
```diff
@@ -0,0 +1,705 @@
+        }
+
+        [ConditionalFact(Helpers.GdiplusIsAvailable)]
+        public void Metafile_String_FileDoesNotExists()
```
Dupe of Ctor_InvalidPath_ThrowsExternalException

mellinoe commented on src/System.Drawing.Common/tests/mono/System.Imaging/MetafileTest.cs:
```diff
@@ -0,0 +1,705 @@
+        }
+
+        [ConditionalFact(Helpers.GdiplusIsAvailable)]
+        public void GetMetafileHeader_Bitmap()
```
Duplicate of Ctor_Bitmap_ThrowsExternalException -- if this is passing then we can enable that other one.

mellinoe commented on src/System.Drawing.Common/tests/mono/System.Imaging/MetafileTest.cs:
```diff
@@ -0,0 +1,705 @@
+
+            Assert.Equal(0, header.Bounds.X);
+            Assert.Equal(0, header.Bounds.Y);
+#if false
```
delete

mellinoe commented on src/System.Drawing.Common/tests/mono/System.Imaging/MetafileTest.cs:
```diff
@@ -0,0 +1,705 @@
+        }
+
+        [ConditionalFact(Helpers.GdiplusIsAvailable)]
+        public void Static_GetMetafileHeader_Filename_Null()
```
Dupe of Static_GetMetafileHeader_NullString_ThrowsArgumentNullException

mellinoe commented on src/System.Drawing.Common/tests/mono/System.Imaging/MetafileTest.cs:
```diff
@@ -0,0 +1,705 @@
+        }
+
+        [ConditionalFact(Helpers.GdiplusIsAvailable)]
+        public void Static_GetMetafileHeader_Stream_Null()
```
Dupe of Static_GetMetafileHeader_NullStream_ThrowsNullReferenceException

mellinoe commented on src/System.Drawing.Common/tests/mono/System.Imaging/MetafileTest.cs:
```diff
@@ -0,0 +1,705 @@
+        }
+
+        [ConditionalFact(Helpers.GdiplusIsAvailable)]
+        public void Static_GetMetafileHeader_Stream()
```
Dupe of Static_GetMetafileHeader_Stream_ReturnsExpected, I think.

mellinoe commented on src/System.Drawing.Common/tests/mono/System.Imaging/MetafileTest.cs:
```diff
@@ -0,0 +1,705 @@
+    public class MetafileFulltrustTest
+    {
+
+        private Font test_font;
```
This field is only used in one test case... Might as well just inline it and properly Dispose of it.

mellinoe commented on src/System.Drawing.Common/tests/mono/System.Imaging/MetafileTest.cs:
```diff
@@ -0,0 +1,705 @@
+        }
+
+        [ConditionalFact(Helpers.GdiplusIsAvailable)]
+        public void Static_GetMetafileHeader_IntPtr_Zero()
```
Duplicate of Static_GetMetafileHeader_IntPtr_ThrowsArgumentException

mellinoe commented on src/System.Drawing.Common/tests/mono/System.Imaging/MetafileTest.cs:
```diff
@@ -0,0 +1,705 @@
+        }
+
+        [ConditionalFact(Helpers.GdiplusIsAvailable)]
+        public void Static_GetMetafileHeader_IntPtr()
```
I think this is covered by  Static_GetMetafileHeader_IntPtr_ThrowsArgumentException and GetHenhmetafile_ReturnsExpected

mellinoe commented on src/System.Drawing.Common/tests/mono/System.Imaging/MetafileTest.cs:
```diff
@@ -0,0 +1,705 @@
+        }
+
+        [ConditionalFact(Helpers.GdiplusIsAvailable)]
+        public void Metafile_IntPtrBool_Zero()
```
Dupe of Ctor_IntPtrZero_ThrowsArgumentException -- reenable it if it's working now.

mellinoe commented on src/System.Drawing.Common/tests/mono/System.Imaging/MetafileTest.cs:
```diff
@@ -0,0 +1,705 @@
+        }
+
+        [ConditionalFact(Helpers.GdiplusIsAvailable)]
+        public void Metafile_IntPtrEmfType_Zero()
```
Dupe of Ctor_ZeroPointerEmfType_ThrowsArgumentException

mellinoe commented on src/System.Drawing.Common/tests/mono/System.Imaging/MetafileTest.cs:
```diff
@@ -0,0 +1,705 @@
+        }
+
+        [ConditionalFact(Helpers.GdiplusIsAvailable)]
+        public void Metafile_IntPtrRectangle_Zero()
```
Covered by Ctor_IntPtrZeroI_ThrowsArgumentException, I think.

mellinoe commented on src/System.Drawing.Common/tests/mono/System.Imaging/MetafileTest.cs:
```diff
@@ -0,0 +1,705 @@
+        }
+
+        [ConditionalFact(Helpers.GdiplusIsAvailable)]
+        public void Metafile_IntPtrRectangleF_Zero()
```
Also covered by Ctor_IntPtrZeroI_ThrowsArgumentException

mellinoe commented on src/System.Drawing.Common/tests/mono/System.Imaging/MetaHeaderTest.cs:
```diff
@@ -0,0 +1,96 @@
+    {
+
+        [ConditionalFact(Helpers.GdiplusIsAvailable)]
+        public void DefaultValues()
```
Dupe of Ctor_Default

mellinoe commented on src/System.Drawing.Common/tests/mono/System.Imaging/MetaHeaderTest.cs:
```diff
@@ -0,0 +1,96 @@
+        }
+
+        [ConditionalFact(Helpers.GdiplusIsAvailable)]
+        public void Min()
```
Covered by ShortProperties_SetValues_ReturnsExpected

mellinoe commented on src/System.Drawing.Common/tests/mono/System.Imaging/MetaHeaderTest.cs:
```diff
@@ -0,0 +1,96 @@
+        }
+
+        [ConditionalFact(Helpers.GdiplusIsAvailable)]
+        public void Max()
```
Also covered by ShortProperties_SetValues_ReturnsExpected

mellinoe commented on src/System.Drawing.Common/tests/mono/System.Imaging/MetaHeaderTest.cs:
```diff
@@ -0,0 +1,96 @@
+namespace MonoTests.System.Drawing.Imaging
+{
+
+    public class MetaHeaderTest
```
I think we are good on MetaHeader tests -- everything here is covered in the existing test class.

mellinoe commented:
@KostaVlev Do you want to take a look for potential duplicates, since you added our `Metafile` tests?

qmfrederik commented:
@mellinoe Removed the dupes, re-enabled the original tests, and rationalized the `using` statements a bit.

KostaVlev commented on src/System.Drawing.Common/tests/mono/System.Imaging/MetafileTest.cs:
```diff
@@ -0,0 +1,705 @@
+
+        [ConditionalFact(Helpers.GdiplusIsAvailable)]
+        public void Metafile_String_Empty()
+        {
```
Duplicates `Ctor_InvalidPath_ThrowsArgumentException`

safern commented on src/System.Drawing.Common/tests/mono/System.Imaging/MetafileTest.cs:
```diff
@@ -0,0 +1,576 @@
+            Assert.Equal(0, header.Bounds.Y);
+
+            try
+            {
```
This whole try, catch block and be just simplified to: `Assert.Throws<ArgumentException>(() => header.WmfHeader);`

mellinoe commented on src/System.Drawing.Common/tests/mono/System.Imaging/MetafileTest.cs:
```diff
@@ -613,6 +499,15 @@ public void CreateFilename_MultipleGraphics_EmfPlusOnly()
+            {
+                test_font = new Font(FontFamily.GenericMonospace, 12);
+            }
+            catch (ArgumentException)
```
Should we fail the test or exit out if we can't create the font? Seems like we could be in for a null-ref or some other problem if we continue on after an exception.

qmfrederik replied:
Makes sens. I guess the try/catch was there historically because this code was in the constructor; catching the exception would let the other tests succeed.

I removed the try/catch alltogether - if the font is not available, this test will now fail with an ArgumentException instead of a NullReferenceException.

mellinoe commented:
The latest version looks good to me. I left one more comment, but it's about an existing problem in the test case.

safern commented on src/System.Drawing.Common/tests/mono/System.Imaging/MetafileTest.cs:
```diff
@@ -0,0 +1,552 @@
+                Check_MetaHeader_WmfPlaceable(mh1);
+
+                MetaHeader mh2 = mf.GetMetafileHeader().WmfHeader;
+                Assert.False(Object.ReferenceEquals(mh1, mh2));
```
You can instead use: `Assert.NotSame(mh1, mh2);` in all the places where you are doing `Assert.False(Object.ReferenceEquals(mh1, mh2));`

https://github.com/xunit/assert.xunit/blob/master/IdentityAsserts.cs#L18

qmfrederik commented:
Cool, I addressed all comments - let's see how CI fares. Note that one product change is embedded in Metafile.cs on Unix. The code now mirrors the Windows code.

safern commented on src/System.Drawing.Common/tests/mono/System.Imaging/MetafileTest.cs:
```diff
@@ -0,0 +1,552 @@
+                    size = new FileInfo(filename).Length;
+                    Assert.Equal(0, size);
+                }
+                // FIXME / doesn't work on mono yet
```
Can we delete these comments?

safern commented on src/System.Drawing.Common/tests/mono/System.Imaging/MetafileTest.cs:
```diff
@@ -0,0 +1,552 @@
+                //				Assert.True (size > 0);
+                if (!single)
+                {
+                    // can we append stuff ?
```
Ditto.

safern commented on src/System.Drawing.Common/tests/mono/System.Imaging/MetafileTest.cs:
```diff
@@ -0,0 +1,552 @@
+                    using (Graphics g = Graphics.FromImage(mf))
+                    {
+                        g.DrawRectangle(Pens.Azure, 10, 10, 80, 80);
+                        // happily no :)
```
Ditto

safern commented on src/System.Drawing.Common/tests/mono/System.Imaging/MetafileTest.cs:
```diff
@@ -0,0 +1,552 @@
+
+        private void CreateFilename(EmfType type, bool single)
+        {
+            string name = String.Format("{0}-{1}.emf", type, single ? "Single" : "Multiple");
```
Nit: `String` => `string`

mellinoe commented:
> Note that one product change is embedded in Metafile.cs on Unix.

Thanks for pointing that out. It looks correct and is the same pattern we use elsewhere.

safern commented on src/System.Drawing.Common/tests/mono/System.Imaging/MetafileTest.cs:
```diff
@@ -0,0 +1,548 @@
+                Check_MetafileHeader_WmfPlaceable(header1);
+
+                MetafileHeader header2 = mf.GetMetafileHeader();
+                Assert.NotSame(header1, header2));
```
This guy has an extra parenthesis, will cause compilation errors :)

safern commented on src/System.Drawing.Common/tests/mono/System.Imaging/MetafileTest.cs:
```diff
@@ -0,0 +1,548 @@
+                Check_MetaHeader_WmfPlaceable(mh1);
+
+                MetaHeader mh2 = mf.GetMetafileHeader().WmfHeader;
+                Assert.NotSame(mh1, mh2));
```
Ditto.

safern commented on src/System.Drawing.Common/tests/mono/System.Imaging/MetafileTest.cs:
```diff
@@ -0,0 +1,548 @@
+                Check_MetaHeader_WmfPlaceable(mh1);
+
+                MetaHeader mh2 = mf.GetMetafileHeader().WmfHeader;
+                Assert.NotSame(mh1, mh2));
```
Ditto.

safern commented on src/System.Drawing.Common/tests/mono/System.Imaging/MetafileTest.cs:
```diff
@@ -0,0 +1,548 @@
+                Check_MetaHeader_WmfPlaceable(mh1);
+
+                MetaHeader mh2 = mf.GetMetafileHeader().WmfHeader;
+                Assert.NotSame(mh1, mh2));
```
Ditto.

KostaVlev commented on src/System.Drawing.Common/tests/mono/System.Imaging/MetafileTest.cs:
```diff
@@ -0,0 +1,705 @@
+
+        [ConditionalFact(Helpers.GdiplusIsAvailable)]
+        public void CreateFilename_MultipleGraphics_EmfPlusOnly()
+        {
```
Tests with "CreateFilename_" prefix seems duplicates `Ctor_StringIntPtrEmfType_Success()`,

qmfrederik replied:
Hmm, those tests fail on Unix though. Let me check.

mellinoe commented on src/System.Drawing.Common/tests/mono/System.Imaging/MetafileTest.cs:
```diff
@@ -0,0 +1,705 @@
+
+        [ConditionalFact(Helpers.GdiplusIsAvailable)]
+        public void Static_GetMetafileHeader_Filename()
+        {
```
Good catch -- we should enable the other test since it's disabled right now.

mellinoe commented on src/System.Drawing.Common/tests/mono/System.Imaging/MetafileTest.cs:
```diff
@@ -0,0 +1,705 @@
+
+        [ConditionalFact(Helpers.GdiplusIsAvailable)]
+        public void Metafile_IntPtrEmfType_Invalid()
+        {
```
This one is disabled too.

mellinoe commented on src/System.Drawing.Common/tests/mono/System.Imaging/MetafileTest.cs:
```diff
@@ -0,0 +1,705 @@
+
+        [ConditionalFact(Helpers.GdiplusIsAvailable)]
+        public void Metafile_IntPtrEmfType_EmfOnly()
+        {
```
Ditto about being disabled. The other two are this same method.

qmfrederik replied:
fyi, they were disabled because they fail because they include this line:
```
var bufferGraphics = Graphics.FromHwndInternal(IntPtr.Zero)
```

This will call into `System.Drawing.LibX11Functions.XOpenDisplay(IntPtr display)` which takes a dependency on X11, which is not enabled on the build servers.

I'm about to replace this by:

```
using (Bitmap bmp = new Bitmap(10, 10, PixelFormat.Format32bppArgb))
using (Graphics g = Graphics.FromImage(bmp))
```

which should work on headless Linux as well.

Increasing coverage, one test a time ;-)

mellinoe replied:
Thanks for fixing that!

qmfrederik commented:
@KostaVlev @safern @mellinoe Thanks for the feedback! I think I addressed most of the comments. 

The PR review starts to look like a real-time treasure hunt - let's see how many duplicates we can find. Exciting ;-)

qmfrederik commented:
As a side-effect, I re-enabled a lot of tests which have been disabled on Unix. They were disabled because they fail because they include this line:
```
var bufferGraphics = Graphics.FromHwndInternal(IntPtr.Zero)
```

This will call into `System.Drawing.LibX11Functions.XOpenDisplay(IntPtr display)` which takes a dependency on X11, which is not enabled on the build servers.

This can be fixed by using this code:
```
using (Bitmap bmp = new Bitmap(10, 10, PixelFormat.Format32bppArgb))
using (Graphics g = Graphics.FromImage(bmp))
```

(repeating the comment here for visibility, because this PR re-enables quite a lot of tests on Unix.)

mellinoe commented:
@qmfrederik Thanks, that is a great improvement. It seems like using a simple bitmap is much more appropriate for these tests.

qmfrederik commented:
There are some exception differences between Linux & Windows that cause CI to fail. They show up now because we re-enabled a bunch of Metafile tests.

I'll have a look at it later.

mmitche commented:
@dotnet-bot test this please (something odd happened, investigating)

mmitche commented:
@dotnet-bot test this please (upgrade done)

qmfrederik commented:
@KostaVlev @safern @mellinoe So it turns out re-enabling all of the Metafile tests on Unix is a bit more work that I had expected - there seems to be a lot of subtle behavior encoded in those tests.

I've split those change off to a separate PR.

mellinoe commented:
@qmfrederik Thanks for the clarification -- I will merge this once CI is green.
